src/repositories/updater.rs

```diff
@@ -0,0 +1,354 @@
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "RepositoryStatsUpdater {{ {:?} }}",
+            self.updaters.iter().map(|u| u.host()).collect::<Vec<_>>()
+        )
+    }
+}
+
+impl RepositoryStatsUpdater {
+    pub fn new(config: &Config) -> Self {
+        let mut updaters: Vec<Box<dyn RepositoryForge + Send + Sync>> = Vec::with_capacity(3);
+        if let Ok(Some(updater)) = GitHub::new(&config) {
+            updaters.push(Box::new(updater));
+        }
+        if let Ok(Some(updater)) = GitLab::new("gitlab.com", &config.gitlab_accesstoken) {
+            updaters.push(Box::new(updater));
+        }
+        if let Ok(Some(updater)) = GitLab::new("gitlab.freedesktop.org", &None) {
+            updaters.push(Box::new(updater));
+        }
+        Self { updaters }
+    }
+
+    pub(crate) fn load_repository(
+        &self,
+        conn: &mut Client,
+        metadata: &MetadataPackage,
+    ) -> Result<Option<i32>> {
+        let url = match &metadata.repository {
+            Some(url) => url,
+            None => {
+                debug!("did not collect stats as no repository URL was present");
+                return Ok(None);
+            }
+        };
+        self.load_repository_inner(conn, url)
+    }
+
+    fn load_repository_inner(&self, conn: &mut Client, url: &str) -> Result<Option<i32>> {
+        let name = match repository_name(url) {
+            Some(name) => name,
+            None => return Ok(None),
+        };
+
+        // Avoid querying the APIs for repositories we already loaded.
+        if let Some(row) = conn.query_opt(
+            "SELECT id FROM repositories WHERE name = $1 AND host = $2 LIMIT 1;",
+            &[&format!("{}/{}", name.owner, name.repo), &name.host],
+        )? {
+            return Ok(Some(row.get("id")));
+        }
+        for updater in self.updaters.iter().filter(|u| u.host() == name.host) {
+            let res = match updater.fetch_repository(&name) {
+                Ok(Some(repo)) => Self::store_repository(conn, updater.host(), repo),
+                Err(err) => {
+                    warn!("failed to collect `{}` stats: {}", updater.host(), err);
+                    return Ok(None);
+                }
+                _ => continue,
+            };
+            return match res {
+                Ok(repo_id) => Ok(Some(repo_id)),
+                Err(err) => {
+                    warn!("failed to collect `{}` stats: {}", updater.host(), err);
+                    Ok(None)
+                }
+            };
+        }
+        // It means that none of our updaters have a matching host.
+        Ok(None)
+    }
+
+    pub fn start_crons(config: Arc<Config>, pool: Pool) -> Result<()> {
+        let instance = Self::new(&config);
+        cron(
+            "repositories stats updater",
+            Duration::from_secs(60 * 60),
+            move || instance.update_all_crates(&pool),
+        )?;
```

Comment:
Why was this moved inside the updater? I think it should still be defined in the daemon.

Reply:
Ah right, time to put it back!